### PR TITLE
Catch bad `name` and raise a `ValueError`

### PR DESCRIPTION
**Pull request by sontek**
This is in response to #1387, I think it helps with the problem but will need more information.

Prior to this:

```
Traceback (most recent call last):
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/channel.py", line 337, in service
    task.service()
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/task.py", line 173, in service
    self.execute()
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/task.py", line 392, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/router.py", line 242, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/router.py", line 216, in invoke_subrequest
    request._set_extensions(extensions)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/util.py", line 114, in _set_extensions
    setattr(self, name, method)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf1' in position 5: ordinal not in range(128)
```

on Py3:

```
Traceback (most recent call last):
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/tests/test_config/test_factories.py", line 133, in test_add_request_method_with_text_type_name
    self.assertRaises(ValueError, set_extensions)
  File "/usr/lib/python3.4/unittest/case.py", line 701, in assertRaises
    return context.handle('assertRaises', callableObj, args, kwargs)
  File "/usr/lib/python3.4/unittest/case.py", line 161, in handle
    callable_obj(*args, **kwargs)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/tests/test_config/test_factories.py", line 131, in set_extensions
    inst._set_extensions(exts)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/util.py", line 115, in _set_extensions
    setattr(self, name, method)
TypeError: attribute name must be string, not 'bytes'
```

After:

```
Traceback (most recent call last):
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/channel.py", line 337, in service
    task.service()
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/task.py", line 173, in service
    self.execute()
  File "/home/sontek/venvs/pyramid/local/lib/python2.7/site-packages/waitress/task.py", line 392, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/router.py", line 242, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/router.py", line 216, in invoke_subrequest
    request._set_extensions(extensions)
  File "/home/sontek/venvs/pyramid/src/pyramid/pyramid/util.py", line 121, in _set_extensions
    raise ValueError(msg % name)
ValueError: `name="La Pe\xf1a"` is invalid. `name` must be ascii because it is used on __name__ of the method

```

### CODE REVIEW

**Comment by bertjwregeer** [comment]
LGTM.

File: pyramid/util.py [hunk 1]
```diff
@@ -111,7 +111,15 @@ def _set_properties(self, properties):
...
         for name, fn in iteritems_(extensions.methods):
             method = fn.__get__(self, self.__class__)
-            setattr(self, name, method)
+            try:
+                setattr(self, name, method)
+            except (UnicodeEncodeError, TypeError):
+                msg = (
+                    '`name="%s"` is invalid. `name` must be ascii because it is '
+                    'used on __name__ of the method'
+                )
```
Comment (mmerickel) [inline pyramid/util.py]: Missing string interpolation for the `%s` here.

**Comment by mmerickel** [comment]
I think we should try to cast the value to an ascii-native string via something like `name = p.compat.native_(name, 'ascii')`.

**Comment by mmerickel** [comment]
Ideally I'd also like to see this be an error at config-time rather than request-time.

**Comment by bertjwregeer** [comment]
What's left to be done on this PR? Just that string interpolation that is missing? Or do we want to add it to check at config time rather than at runtime?

**Comment by mmerickel** [comment]
1) Fix interpolation. It's broken.
2) I think a config-time check is worthwhile here if we're going through this trouble to help out.

File: pyramid/util.py (hunk 1 above)
Comment (sontek) [inline pyramid/util.py]: I interpolate it on the very next line

**Comment by sontek** [comment]
@mmerickel Pushed the check to configuration time instead of runtime

**Comment by bertjwregeer** [comment]
@sontek testing broke on Python 3.2 due to u''.

File: pyramid/tests/test_util.py [hunk 2]
```diff
@@ -619,7 +634,30 @@ def test___str__(self):
                          "Line 0 of file filename:\n       linerepr  ")
 
 
+class TestCallableName(unittest.TestCase):
+    def test_valid_ascii(self):
+        from pyramid.util import get_callable_name
+        name = u'hello world'
```
Comment (bertjwregeer) [inline pyramid/tests/test_util.py]: u'' is not valid on Python 3.2 and breaks tests.

Comment (sontek) [inline pyramid/tests/test_util.py]: Yeah, I refuse to install that shitty version of python on my system and always forget :P

**Comment by mmerickel** [comment]
Whee, thanks!

**Pull request merged**